# OpenWeather API Testing

--------------

This notebook's purpose is to test the [One Call API 3.0](https://openweathermap.org/api/one-call-3) from Open Weather API

In [7]:
# import statements

import requests
import os
import datetime as dt
from dotenv import load_dotenv, find_dotenv

In [8]:
# import local environmental variables (so API Keys are hidden from github repo)

load_dotenv(find_dotenv(), override=True)

True

In [11]:
API_KEY = os.getenv('BLINKMETRICS_OPENWEATHER_API_KEY')

To convert from city names to lat-lon we will use the [Geocoding API](https://openweathermap.org/api/geocoding-api)

In [14]:
# testing parameters for the geocoding API

city_name = 'Beirut'
country_code = 'LB'
limit = 2

geo_code_url = f'http://api.openweathermap.org/geo/1.0/direct?q={city_name},{country_code}&limit={limit}&appid={API_KEY}'

In [19]:
# calling the api and seeing the status code (if the call worked or not)

geo_response = requests.get(geo_code_url)
geo_response.status_code

200

In [25]:
# getting the result

geo_result = geo_response.json()
geo_result

[{'name': 'Beirut',
  'local_names': {'ascii': 'Beirut',
   'oc': 'Beirot',
   'bn': 'বৈরুত',
   'et': 'Beirut',
   'lv': 'Beirūta',
   'ru': 'Бейрут',
   'es': 'Beirut',
   'br': 'Beirout',
   'bs': 'Bejrut',
   'fr': 'Beyrouth',
   'fy': 'Beirût',
   'cs': 'Bejrút',
   'ku': 'Bêrût',
   'ur': 'بیروت',
   'ar': 'بيروت',
   'is': 'Beirút',
   'mk': 'Бејрут',
   'am': 'ቤይሩት',
   'lt': 'Beirutas',
   'kk': 'Бейрут',
   'mr': 'बैरूत',
   'af': 'Beiroet',
   'he': 'ביירות',
   'pl': 'Bejrut',
   'fi': 'Beirut',
   'sv': 'Beirut',
   'el': 'Βηρυτός',
   'tg': 'Бейрут',
   'hu': 'Bejrút',
   'ka': 'ბეირუთი',
   'sr': 'Бејрут',
   'bo': 'པེ་རུ་ཐེ།',
   'yi': 'ביירוט',
   'th': 'เบรุต',
   'fa': 'بیروت',
   'pt': 'Beirute',
   'it': 'Beirut',
   'de': 'Beirut',
   'uk': 'Бейрут',
   'ug': 'بېيرۇت',
   'ml': 'ബെയ്റൂത്ത്',
   'ta': 'பெய்ரூத்',
   'sk': 'Bejrút',
   'be': 'Бейрут',
   'my': 'ဘေရွတ်မြို့',
   'eu': 'Beirut',
   'eo': 'Bejruto',
   'vo': 'Bäyrut',
   'feature_name': 'Beirut',
   'd

In [27]:
# getting the lat and lon values:

lat = geo_result[0]['lat']
lon = geo_result[0]['lon']

print((lat,lon))

(33.8959203, 35.47843)


Now to try out the One Call API

In [72]:
# testing some parameters

lat = lat
lon = lon
units = 'metric'

oc_url = f'https://api.openweathermap.org/data/3.0/onecall?lat={lat}&lon={lon}&units={units}&appid={API_KEY}'

In [73]:
# calling the api and seeing the status code (if the call worked or not)

oc_response = requests.get(oc_url)
oc_response.status_code

401

Since it's returning 401, I need to figure out what is happening since we are getting an error.

In [77]:
# printing the error message (excluding the url so that the API KEY does not show up)

try:
    oc_response.raise_for_status()
    
except Exception as e:
    print(str(e).split(': https')[0])

401 Client Error: Unauthorized for url


The request to the API is failing due to an authorization issue, let us see how to fix it.

In [64]:
# trying the api call for current weather

lat = lat
lon = lon

curr_url = f'https://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}&appid={API_KEY}'

In [65]:
curr_response = requests.get(curr_url)
curr_response.status_code

200

In [66]:
curr_response.json()

{'coord': {'lon': 35.4809, 'lat': 33.9016},
 'weather': [{'id': 801,
   'main': 'Clouds',
   'description': 'few clouds',
   'icon': '02d'}],
 'base': 'stations',
 'main': {'temp': 301.96,
  'feels_like': 303.58,
  'temp_min': 301.54,
  'temp_max': 301.96,
  'pressure': 1013,
  'humidity': 58,
  'sea_level': 1013,
  'grnd_level': 1006},
 'visibility': 8000,
 'wind': {'speed': 3.09, 'deg': 250},
 'clouds': {'all': 20},
 'dt': 1728988568,
 'sys': {'type': 1,
  'id': 7532,
  'country': 'LB',
  'sunrise': 1728963780,
  'sunset': 1729004656},
 'timezone': 10800,
 'id': 268743,
 'name': 'Ra’s Bayrūt',
 'cod': 200}

In [67]:
# trying the api call for the 3 hours 5 days forecasting

lat = lat
lon = lon

forecast_url = f'https://api.openweathermap.org/data/2.5/forecast?lat={lat}&lon={lon}&appid={API_KEY}'

In [69]:
forecast_response = requests.get(forecast_url)
forecast_response.status_code

200

In [70]:
forecast_response.json()

{'cod': '200',
 'message': 0,
 'cnt': 40,
 'list': [{'dt': 1728993600,
   'main': {'temp': 301.76,
    'feels_like': 303.27,
    'temp_min': 298.8,
    'temp_max': 301.76,
    'pressure': 1013,
    'sea_level': 1013,
    'grnd_level': 1006,
    'humidity': 58,
    'temp_kf': 2.96},
   'weather': [{'id': 801,
     'main': 'Clouds',
     'description': 'few clouds',
     'icon': '02d'}],
   'clouds': {'all': 20},
   'wind': {'speed': 1.63, 'deg': 263, 'gust': 1.56},
   'visibility': 10000,
   'pop': 0,
   'sys': {'pod': 'd'},
   'dt_txt': '2024-10-15 12:00:00'},
  {'dt': 1729004400,
   'main': {'temp': 300.69,
    'feels_like': 302.04,
    'temp_min': 298.54,
    'temp_max': 300.69,
    'pressure': 1013,
    'sea_level': 1013,
    'grnd_level': 1006,
    'humidity': 61,
    'temp_kf': 2.15},
   'weather': [{'id': 801,
     'main': 'Clouds',
     'description': 'few clouds',
     'icon': '02d'}],
   'clouds': {'all': 13},
   'wind': {'speed': 1.68, 'deg': 338, 'gust': 1.44},
   'visibilit

The problem was that the One Call API free subscription does not use the same URL for the paid subscription, so i had to go and get the urls for the available features of the free subscription, ir: the url for the [Current Weather API](https://openweathermap.org/current) and [3 Hour Forecast 5 Days API](https://openweathermap.org/forecast5)

So, when proceeding forward to using the free plan of the One Call API from Open Weather API we will use the Current Weather API and 3 Hour Forecast 5 Days API!